### DISCUSSION TITLE
Misleading warning: "Object of type <ParentClass> not in session, add operation along 'ChildClass.relationship' will not proceed

### DISCUSSION THREAD

**Discussion by seanh** [sqlalchemy/sqlalchemy#12049] [Usage Questions] [answered]
Hi,

I'm getting this warning which I think is a false-positive, and I think SQLAlchemy may want to reconsider emitting the warning as it may be confusing and inconvenient (requiring some code contortions to avoid triggering false-positive warnings).

Alternatively if I've misunderstood what's going on please enlighten me!

The warning:

```
sqlalchemy.exc.SAWarning: Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
```

I'm using the "association object" pattern from this page: https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object

Here's the relevant parts of my classes:

```python
class User(Base):
    __tablename__ = "user"
    id = Column(Integer, autoincrement=True, primary_key=True)
    memberships = relationship("GroupMembership", back_populates="user")

class Group(Base):
    __tablename__ = "group"
    id = Column(Integer, autoincrement=True, primary_key=True)
    memberships = relationship("GroupMembership", back_populates="group")

class GroupMembership(Base):
    __tablename__ = "user_group"
    __table_args__ = (sa.UniqueConstraint("user_id", "group_id"),)
    id = Column(Integer, autoincrement=True, primary_key=True)
    user_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    user = relationship("User", back_populates="memberships")
    group_id = Column(Integer, ForeignKey("group.id", ondelete="cascade"), nullable=False, index=True)
    group = relationship("Group", back_populates="memberships")
    roles = Column(JSONB, server_default=text("""'["member"]'::jsonb"""), nullable=False)
```

As you can see my `GroupMembership` association class has an extra attribute of its own (`roles`), hence why I'm using the association object pattern. But I don't think that's relevant to the issue at hand.

And here's some code that triggers the false-positive warning:

```python
>>> membership = models.GroupMembership(user=user)
>>> group.memberships.append(membership)
SAWarning: Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
```

...where `user` and `group` are `User` and `Group` objects that are already in the SQLAlchemy session.

What I think is happening is that this line:

```python
group.memberships.append(membership)
```

...first evaluates the `group.memberships` relationship which triggers an autoflush. At this point `membership` has not yet been added to the session, but `membership.user` _has_ been added to the session. Hence SQLAlchemy's warning: `Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed`. However, Python will then evaluate the rest of the line (`.append(membership)`) and appending `membership` to `group.memberships` _will_ have the side-effect of adding `membership` to the session.

Hence immediately after this line (and immediately after SQLAlchemy has just said `Object of type <GroupMembership> not in session`) if you query whether `membership` is in the session you'll find that it _is_.

This variation of the line will not trigger the warning because the autoflush is triggered before the `GroupMembership` is initialized:

```python
group.memberships.append(GroupMembership(user=user))
```

This version also avoids the warning:

```python
membership = models.GroupMembership(user_id=user.id)
group.memberships.append(membership)
```

_Assigning_ to `group.memberships` also triggers the same warning:

```python
>>> membership = models.GroupMembership(user=user)
>>> group.memberships = [membership]
SAWarning: Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
```

With assignment the warning is not so easy to avoid. I think for assignments Python evaluates the right-hand-side first so this variation still triggers the warning:

```python
>>> group.memberships = [models.GroupMembership(user=user)]
SAWarning: Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
```

This variation does avoid the warning:

```python
membership = models.GroupMembership(user_id=user.id)
group.memberships = [membership]
```

If you have a handle to the SQLAlchemy session you can of course add your `GroupMembership`'s to it directly and avoid directly mutating the `User.memberships` or `Group.memberships` relationships, this will also avoid any warnings but it can be inconvenient because you have to pass the `session` object around everywhere:

```python
session.add(GroupMembership(user=user, group=group))
```

In all of the above examples the `GroupMembership` _was_ added to the session immediately after SQLAlchemy's warning that it would not be. This had me confused for quite a few hours, and I was further confused by the warning being either triggered or not-triggered by seemingly very similar code examples like the ones above.

And as you can see I'm having to contort my code a bit to avoid triggering the warning. It would be nice to be able to do this sort of thing without triggering false-positive warnings:

```python
membership = GroupMembership(user=user)
group.memberships.append(membership)
```

One last note, this won't work:

```python
membership = GroupMembership()
group.memberships.append(membership)
user.memberships.append(membership)
```

...the last line triggers a `NotNullViolation` because it tries to flush `membership` while `membership.user` is still `None`.

For reference (and the full code) this is the PR in my project where I ran into these warnings: https://github.com/hypothesis/h/pull/9047

Thanks!

**Comment by zzzeek** [comment, marked as answer]
>  SQLAlchemy may want to reconsider emitting the warning 

This warning isn't just a thing that SQLAlchemy "sees" but otherwise has no effect on anything.  It means the unit of work is making a choice about objects that it has encountered which are in a place that they shouldn't be, namely, in a collection that is maintained by a relationship yet they aren't in the Session.    If it didnt emit the warning (which you can as always silence with the [warnings filter](https://docs.python.org/3/library/warnings.html) ) it still is making a choice about objects which it's encountering inside of a collection.

> ..first evaluates the group.memberships relationship which triggers an autoflush. At this point membership has not yet been added to the session, but membership.user has been added to the session. Hence SQLAlchemy's warning: Object of type <GroupMembership> not in session, add operation along 'User.memberships' will not proceed. However, Python will then evaluate the rest of the line (.append(membership)) and appending membership to group.memberships will have the side-effect of adding membership to the session.

that's correct, the warning does indicate that it's happening in autoflush, and that flush happens before "membership" is part of the session.   this is because [sqlalchemy no longer reverse-cascades objects into the session](https://docs.sqlalchemy.org/en/14/orm/cascades.html#backref-cascade) so when you did `GroupMembership(user=user)`, that no longer adds the object to the session.   but it does add it to `user.memberships`.  It's there, and the unit of work has no idea that you'd like this change to the collection to be ignored.  it's not enthusiastic about ignoring an object in a collection that it normally would be flushing.   so it warns.

prior to the change involving reverse cascade, the code fails the other way, like this:

```
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(psycopg2.errors.NotNullViolation) null value in column "group_id" of relation "user_group" violates not-null constraint
DETAIL:  Failing row contains (1, 1, null, ["member"]).

```

so it fails both ways.   the solution to the issue is the same in both cases which is to do a no_autoflush when you are constructing association objects against other objects that might be triggering autoflush:

```py
    with sess.no_autoflush:
        membership = GroupMembership(user=user)
        group.memberships.append(membership)

```

**Reply by seanh** [reply]
Thanks! It's good to have confirmation that my understanding is correct and there's nothing actually wrong with my code. And thanks for pointing out `session.no_autoflush`, I didn't know about that and it ended up being the best solution in a few places in my PR.

**Reply by CaselIT** [reply]
I will update the warning to mention the no_autoflush option.

**Reply by CaselIT** [reply]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5560
